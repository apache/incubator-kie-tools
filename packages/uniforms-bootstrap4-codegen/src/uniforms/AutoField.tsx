/*
 * Copyright 2021 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { createAutoField } from "uniforms/es5/createAutoField";
import {
  BoolField,
  CheckBoxGroupField,
  DateField,
  NestField,
  NumField,
  RadioField,
  SelectField,
  TextField,
  UnsupportedField,
} from "./index";

export type AutoFieldProps = Parameters<typeof AutoField>[0];

const AutoField = createAutoField((props) => {
  if (props.allowedValues) {
    if (props.checkboxes) {
      return props.fieldType !== Array ? RadioField : CheckBoxGroupField;
    }
    return SelectField;
  }

  switch (props.fieldType) {
    /*
    TODO: implement array support
    case Array:
      return  ListField;*/
    case Boolean:
      return BoolField;
    case Date:
      return DateField;
    case Number:
      return NumField;
    case Object:
      return NestField;
    case String:
      return TextField;
  }

  console.log(`Unsupported field type: ${props.fieldType.name}`);
  return UnsupportedField;
});

export default AutoField;
