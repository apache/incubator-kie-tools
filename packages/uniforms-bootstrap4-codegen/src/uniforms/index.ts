/*
 * Copyright 2021 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

export { default as AutoField } from "./AutoField";
export { default as AutoFields } from "./AutoFields";
export { default as AutoForm } from "./AutoForm";
export { default as BoolField } from "./BoolField";
export { default as CheckBoxGroupField } from "./CheckBoxGroupField";
export { default as DateField } from "./DateField";
export { default as NestField } from "./NestField";
export { default as NumField } from "./NumField";
export { default as RadioField } from "./RadioField";
export { default as SelectField } from "./SelectField";
export { default as TextField } from "./TextField";
export { default as UnsupportedField } from "./UnsupportedField";

export { renderForm } from "./renderForm";
