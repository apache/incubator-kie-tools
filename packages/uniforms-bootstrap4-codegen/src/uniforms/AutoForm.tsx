/*
 * Copyright 2021 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as React from "react";
import escape from "lodash/escape";
import { Bridge } from "uniforms";
import { CodeGenElement } from "../api";
import { FORM, renderCodeGenElement } from "./templates/templates";
import { renderFormInputs } from "./rendering/RenderingUtils";

export type AutoFormProps = {
  id: string;
  disabled?: boolean;
  placeholder?: boolean;
  schema: Bridge;
};

const AutoForm: React.FC<AutoFormProps> = (props) => {
  const properties = {
    children: renderFormInputs(props.schema),
  };

  const form: CodeGenElement = renderCodeGenElement(FORM, properties);

  return <>{escape(form.html)}</>;
};

export default AutoForm;
