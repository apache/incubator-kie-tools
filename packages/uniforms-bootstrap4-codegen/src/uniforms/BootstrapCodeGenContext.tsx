/*
 * Copyright 2021 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React, { createContext, useContext } from "react";

import { Bridge, context, Context, randomIds } from "uniforms/es5";

import { FormElement } from "../api";

export interface BootstrapCodeGenContext {
  rendered: FormElement<any>[];
}

export const bootstrapCodeGenContext = createContext<BootstrapCodeGenContext | null>(null);

export const useBootstrapCodegenContext = (): BootstrapCodeGenContext | null => {
  return useContext<BootstrapCodeGenContext | null>(bootstrapCodeGenContext);
};

export const useAddFormElementToBootstrapContext = (formElement: FormElement<any>): void => {
  const ctx = useBootstrapCodegenContext();
  if (!ctx) {
    throw new Error(`'useAddFormElementToBootstrapContext' should be called within a 'bootstrapCodeGenContext'`);
  }
  ctx.rendered.push(formElement);
};

export interface ProviderProps {
  schema: Bridge;
  codegenCtx: BootstrapCodeGenContext;
  uniformsCtx?: Context<any>;
  children: JSX.Element;
}

export const CodeGenContextProvider: React.FC<ProviderProps> = (props) => {
  const ctx: Context<any> = props.uniformsCtx || {
    changed: false,
    changedMap: {},
    error: false,
    model: {},
    name: [],
    onChange: (key, value) => {},
    onSubmit: (event) => {},
    randomId: randomIds(),
    schema: props.schema,
    state: {
      readOnly: false,
      disabled: false,
      label: true,
      placeholder: true,
      showInlineError: true,
    },
    submitted: false,
    submitting: false,
    validating: false,
  };

  return (
    <bootstrapCodeGenContext.Provider value={props.codegenCtx}>
      <context.Provider value={ctx}>{props.children}</context.Provider>
    </bootstrapCodeGenContext.Provider>
  );
};
