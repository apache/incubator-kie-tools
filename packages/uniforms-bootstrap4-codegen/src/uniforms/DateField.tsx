/*
 * Copyright 2021 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as React from "react";
import { connectField, HTMLFieldProps } from "uniforms/es5";
import { DATE, renderCodeGenElement } from "./templates/templates";
import { useAddFormElementToBootstrapContext } from "./BootstrapCodeGenContext";
import { FormInput } from "../api";

export type DateFieldProps = HTMLFieldProps<
  Date,
  HTMLDivElement,
  {
    name: string;
    label: string;
    required: boolean;
    max?: Date;
    min?: Date;
  }
>;

const Date: React.FC<DateFieldProps> = (props: DateFieldProps) => {
  function formatDate(date?: Date) {
    return date?.toISOString().slice(0, -8);
  }

  const properties = {
    id: props.id,
    name: props.name,
    label: props.label,
    disabled: props.disabled ?? false,
    placeholder: props.placeholder,
    autoComplete: props.autoComplete ?? false,
    value: formatDate(props.value),
    max: formatDate(props.max),
    min: formatDate(props.min),
  };

  const element: FormInput = renderCodeGenElement(DATE, properties);
  useAddFormElementToBootstrapContext(element);
  return <>{JSON.stringify(element)}</>;
};

export default connectField(Date);
