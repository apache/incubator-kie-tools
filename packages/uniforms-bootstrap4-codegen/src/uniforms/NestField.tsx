/*
 * Copyright 2021 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React, { useContext } from "react";
import { connectField, context, HTMLFieldProps } from "uniforms/es5";
import { renderNestedInputFragmentWithContext } from "./rendering/RenderingUtils";
import { FormElement, FormInput, FormInputContainer } from "../api";
import { useBootstrapCodegenContext } from "./BootstrapCodeGenContext";
import { NESTED, renderCodeGenElement } from "./templates/templates";

export type NestFieldProps = HTMLFieldProps<object, HTMLDivElement, { itemProps?: object }>;

const Nest: React.FunctionComponent<NestFieldProps> = ({
  id,
  children,
  error,
  errorMessage,
  fields,
  itemProps,
  label,
  name,
  showInlineError,
  disabled,
  ...props
}: NestFieldProps) => {
  const uniformsContext = useContext(context);
  const codegenCtx = useBootstrapCodegenContext();

  const nestedFields: FormElement<any>[] = [];
  if (fields) {
    fields.forEach((field) => {
      const nestedElement = renderNestedInputFragmentWithContext(uniformsContext, field, itemProps, disabled);
      if (nestedElement) {
        nestedFields.push(nestedElement);
      } else {
        console.log(`Cannnot render form field for: '${field}'`);
      }
    });
  }

  const properties = {
    id: id,
    name: name,
    label: label,
    disabled: disabled,
    children: nestedFields,
  };

  const element: FormInputContainer = renderCodeGenElement(NESTED, properties);
  codegenCtx?.rendered.push(element);
  return <>{JSON.stringify(element)}</>;
};

export default connectField(Nest);
