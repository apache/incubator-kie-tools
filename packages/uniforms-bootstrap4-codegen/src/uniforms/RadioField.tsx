/*
 * Copyright 2021 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as React from "react";
import { connectField, HTMLFieldProps } from "uniforms/es5";
import { RADIOGROUP, renderCodeGenElement } from "./templates/templates";
import { useAddFormElementToBootstrapContext } from "./BootstrapCodeGenContext";
import { FormInput } from "../api";

export type RadioFieldProps = HTMLFieldProps<
  string,
  HTMLDivElement,
  {
    name: string;
    label: string;
    transform?: (string?: string) => string;
    allowedValues?: string[];
    required: boolean;
    disabled: boolean;
  }
>;

const Radio = (props: RadioFieldProps) => {
  const options =
    props.allowedValues?.map((option) => {
      return {
        value: option,
        label: props.transform ? props.transform(option) : option,
        checked: props.value === option,
      };
    }) || [];

  const inputProps = {
    id: props.id,
    name: props.name,
    label: props.label,
    disabled: props.disabled,
    options: options,
  };

  const element: FormInput = renderCodeGenElement(RADIOGROUP, inputProps);
  useAddFormElementToBootstrapContext(element);
  return <>{JSON.stringify(element)}</>;
};

export default connectField(Radio);
