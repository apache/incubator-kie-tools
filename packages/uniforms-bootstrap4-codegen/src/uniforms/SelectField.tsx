/*
 * Copyright 2021 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as React from "react";
import { connectField, HTMLFieldProps } from "uniforms/es5";
import { renderCodeGenElement, SELECT } from "./templates/templates";
import { useAddFormElementToBootstrapContext } from "./BootstrapCodeGenContext";
import { FormInput } from "../api";

export type SelectInputProps = HTMLFieldProps<
  string[],
  HTMLDivElement,
  {
    name: string;
    label: string;
    placeHolder: string;
    allowedValues?: string[];
    required: boolean;
    transform?(value: string): string;
  }
>;

const Select: React.FC<SelectInputProps> = (props: SelectInputProps) => {
  const options =
    props.allowedValues?.map((option) => {
      return {
        value: option,
        label: props.transform ? props.transform(option) : option,
        checked: props.value?.includes(option),
      };
    }) || [];

  const inputProps = {
    id: props.id,
    name: props.name,
    label: props.label,
    multiple: props.fieldType === Array,
    placeHolder: props.placeHolder,
    disabled: props.disabled,
    options: options,
    value: props.value,
  };

  const element: FormInput = renderCodeGenElement(SELECT, inputProps);
  useAddFormElementToBootstrapContext(element);
  return <>{JSON.stringify(element)}</>;
};

export default connectField(Select);
