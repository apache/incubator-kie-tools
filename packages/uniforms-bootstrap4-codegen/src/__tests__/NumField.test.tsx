/*
 * Copyright 2021 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as React from "react";
import SimpleSchema from "simpl-schema";
import { renderField } from "./_render";
import { NumField } from "../uniforms";

describe("<NumField> tests", () => {
  it("<NumField> - integer rendering", () => {
    const schema = {
      age: { type: SimpleSchema.Integer },
    };

    const props = {
      id: "id",
      label: "age",
      name: "age",
      min: 10,
      max: 15,
      disabled: false,
    };

    const { container, formElement } = renderField(NumField, props, schema);

    expect(container).toMatchSnapshot();

    expect(formElement.html).toContain(`<label for="${props.id}">${props.label}</label>`);
    expect(formElement.html).toContain(`id="${props.id}"`);
    expect(formElement.html).toContain('type="number"');
    expect(formElement.html).toContain(`name="${props.name}"`);
    expect(formElement.html).toContain(`min="${props.min}"`);
    expect(formElement.html).toContain(`max="${props.max}"`);
    expect(formElement.html).toContain(`step="1"`);
    expect(formElement.html).not.toContain("autoComplete");
    expect(formElement.html).not.toContain("disabled");
    expect(formElement.html).not.toContain("placeholder");

    expect(formElement.ref.binding).toBe(props.name);
  });

  it("<NumField> - integer rendering - no min/max", () => {
    const schema = {
      age: { type: SimpleSchema.Integer },
    };

    const props = {
      id: "id",
      label: "age",
      name: "age",
      onChange: jest.fn(),
    };

    const { formElement } = renderField(NumField, props, schema);

    expect(formElement.html).toContain(`<label for="${props.id}">${props.label}</label>`);
    expect(formElement.html).toContain(`id="${props.id}"`);
    expect(formElement.html).toContain('type="number"');
    expect(formElement.html).toContain(`name="${props.name}"`);
    expect(formElement.html).not.toContain("min=");
    expect(formElement.html).not.toContain("max=");
  });

  it("<NumField> - integer rendering - disabled", () => {
    const schema = {
      age: { type: SimpleSchema.Integer },
    };

    const props = {
      id: "id",
      label: "age",
      name: "age",
      min: 10,
      max: 15,
      disabled: true,
    };

    const { container, formElement } = renderField(NumField, props, schema);

    expect(container).toMatchSnapshot();

    expect(formElement.html).toContain(`<label for="${props.id}">${props.label}</label>`);
    expect(formElement.html).toContain(`id="${props.id}"`);
    expect(formElement.html).toContain('type="number"');
    expect(formElement.html).toContain(`name="${props.name}"`);
    expect(formElement.html).toContain("disabled");
  });

  it("<NumField> - decimal rendering", () => {
    const schema = {
      salary: { type: Number },
    };

    const props = {
      id: "id",
      label: "salary",
      name: "salary",
      min: 400.5,
      max: 900.5,
      disabled: false,
    };

    const { container, formElement } = renderField(NumField, props, schema);

    expect(container).toMatchSnapshot();

    expect(formElement.html).toContain(`<label for="${props.id}">${props.label}</label>`);
    expect(formElement.html).toContain(`id="${props.id}"`);
    expect(formElement.html).toContain('type="number"');
    expect(formElement.html).toContain(`name="${props.name}"`);
    expect(formElement.html).toContain(`min="${props.min}"`);
    expect(formElement.html).toContain(`max="${props.max}"`);
    expect(formElement.html).toContain(`step="0.01"`);
    expect(formElement.html).not.toContain('autoComplete"');
    expect(formElement.html).not.toContain('disabled"');
    expect(formElement.html).not.toContain("placeholder");
  });
});
