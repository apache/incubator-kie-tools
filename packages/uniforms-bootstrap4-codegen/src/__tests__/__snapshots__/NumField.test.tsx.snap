// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NumField> tests <NumField> - decimal rendering 1`] = `
<div>
  {"ref":{"id":"id","binding":"salary"},"html":"&lt;div class=\\"form-group\\"&gt;\\n    &lt;label for=\\"id\\"&gt;salary&lt;/label&gt;\\n    &lt;input\\n    type=\\"number\\"\\n    class=\\"form-control\\"\\n    id=\\"id\\"\\n    name=\\"salary\\"\\n    \\n    \\n    \\n    min=\\"400.5\\"\\n    max=\\"900.5\\"\\n    step=\\"0.01\\"\\n    value=\\"\\"/&gt;\\n&lt;/div&gt;"}
</div>
`;

exports[`<NumField> tests <NumField> - integer rendering - disabled 1`] = `
<div>
  {"ref":{"id":"id","binding":"age"},"html":"&lt;div class=\\"form-group\\"&gt;\\n    &lt;label for=\\"id\\"&gt;age&lt;/label&gt;\\n    &lt;input\\n    type=\\"number\\"\\n    class=\\"form-control\\"\\n    id=\\"id\\"\\n    name=\\"age\\"\\n    \\n    disabled\\n    \\n    min=\\"10\\"\\n    max=\\"15\\"\\n    step=\\"1\\"\\n    value=\\"\\"/&gt;\\n&lt;/div&gt;"}
</div>
`;

exports[`<NumField> tests <NumField> - integer rendering 1`] = `
<div>
  {"ref":{"id":"id","binding":"age"},"html":"&lt;div class=\\"form-group\\"&gt;\\n    &lt;label for=\\"id\\"&gt;age&lt;/label&gt;\\n    &lt;input\\n    type=\\"number\\"\\n    class=\\"form-control\\"\\n    id=\\"id\\"\\n    name=\\"age\\"\\n    \\n    \\n    \\n    min=\\"10\\"\\n    max=\\"15\\"\\n    step=\\"1\\"\\n    value=\\"\\"/&gt;\\n&lt;/div&gt;"}
</div>
`;
